pygsp/_nearest_neighbor.py

```diff
@@ -0,0 +1,247 @@
+# -*- coding: utf-8 -*-
+
+from __future__ import division
+
+import numpy as np
+from scipy import sparse, spatial
+from pygsp import utils
+
+_logger = utils.build_logger(__name__)
+
+def _scipy_pdist(features, metric, order, kind, k, radius, params):
+    if params:
+        raise ValueError('unexpected parameters {}'.format(params))
+    metric = 'cityblock' if metric == 'manhattan' else metric
+    metric = 'chebyshev' if metric == 'max_dist' else metric
+    params = dict(metric=metric)
+    if metric == 'minkowski':
+        params['p'] = order
+    dist = spatial.distance.pdist(features, **params)
+    dist = spatial.distance.squareform(dist)
+    if kind == 'knn':
+        neighbors = np.argsort(dist)[:, :k+1]
+        distances = np.take_along_axis(dist, neighbors, axis=-1)
+    elif kind == 'radius':
+        distances = []
+        neighbors = []
+        for distance in dist:
+            neighbor = np.flatnonzero(distance < radius)
+            neighbors.append(neighbor)
+            distances.append(distance[neighbor])
+    return neighbors, distances
+
+
+def _scipy_kdtree(features, _, order, kind, k, radius, params):
+    if order is None:
+        raise ValueError('invalid metric for scipy-kdtree')
+    eps = params.pop('eps', 0)
+    tree = spatial.KDTree(features, **params)
+    params = dict(p=order, eps=eps)
+    if kind == 'knn':
+        params['k'] = k + 1
+    elif kind == 'radius':
+        params['k'] = None
+        params['distance_upper_bound'] = radius
+    distances, neighbors = tree.query(features, **params)
+    return neighbors, distances
+
+
+def _scipy_ckdtree(features, _, order, kind, k, radius, params):
+    if order is None:
+        raise ValueError('invalid metric for scipy-kdtree')
+    eps = params.pop('eps', 0)
+    tree = spatial.cKDTree(features, **params)
+    params = dict(p=order, eps=eps, n_jobs=-1)
+    if kind == 'knn':
+        params['k'] = k + 1
+    elif kind == 'radius':
+        params['k'] = features.shape[0]  # number of vertices
```

Comment:
I think there is a problem here. This does not scale. 
According to the doc of `scipy.spatial.cKDTree.query`

> If x has shape tuple+(self.m,), then d has shape tuple+(k,).

Which means `tree.query()` is going to return a huge matrix of NxN.

This is actually different from `scipy.spatial.KDTree.query` which, according to the doc, returns :

>  If k is None, then d is an object array of shape tuple, containing lists of distances.

Which, if I understand correctly, means that `tree.query()` is going to return an array of size N containing lists with varying length, depending on the number of neighbours in the epsilon ball.

Anyway, I am testing this locally with the fountain dataset that you made available [here](https://lts2.epfl.ch/research/reproducible-research/graph-based-point-cloud-denoising/) and it seems that constructing a radius NN graph using this branch of pygsp crashes
```
G = graphs.NNGraph(points, kind='radius', radius=0.01, 
                   standardize=False, kernel_width=0.1)
```
whereas using the pip version of pygsp does not.
```
G = graphs.NNGraph(points, NNtype='radius', epsilon=0.01, 
                   rescale=False, sigma=0.1)
```

On my initial experiments, I also used `scipy.spatial.cKDTree.query_ball_point()` which only returns the indices, so distances need to be computed seperately. That might be an alternative maybe?

Reply:
See PR #60 about the use of `scipy.spatial.cKDTree.query_ball_point()`